epstoimg.1
.TH EPSTOIMG "1" "September 2003" "epstoimg 1.0" "User Commands"
.SH NAME
epstoimg \- create bitmap images from PostScript
.SH SYNOPSIS
epstoimg [options]... <filename>
.SH DESCRIPTION
This program uses the Ghostscript interpreter
(http://www.ghostscript.com/) to produce bitmap images from PostScript
language files.  The image shows the area within the bounding box, if
the input file contains one, otherwise it shows the entire page.
Multipage documents are supported.
.SH OPTIONS
.TP
\fB\-?\fR, \fB\-\-help\fR
Causes the program to show an option summary and usage information, then exit.
.TP
\fB\-\-version\fR
Causes the program to show version and license information, then exit.
.TP
\fB\-w\fR, \fB\-\-width=\fR#
.TP
\fB\-h\fR, \fB\-\-height=\fR#
Sets the size of the output image, in pixels.  This will be the total
size of the image including any padding.  If only one is given, the
other is calculated using the EPS bounding box, or the page size if no
bounding box is specified.  If neither is specified, the output image
is sized to have one pixel per PostScript point in the input, plus any
padding.
.TP
\fB\-m\fR, \fB\-\-margin=\fR<dim>
If a bounding box is specified in the input file, it is padded by this
length on all sides before the file is rendered.  This value may be
negative, so long as the resulting box is nondegenerate.

The length is given as a number followed by
"in", "cm", "mm", "pt", "ft", or "m" to indicate the units.  A number
given without units is assumed to be a distance in PostScript points
(1/72-inch).
.TP
\fB\-p\fR, \fB\-\-padding=\fR#
After rendering, the output image can be padded on all sides by a border
of white pixels.  The default is zero pixels (no padding).
.TP
\fB\-b\fR, \fB\-\-black\fR
This causes the \fB\-p\fR option to pad with black rather than white.
.TP
\fB\-r\fR, \fB\-\-rotation=\fR#
This rotates the output image by the given (integral) number of
counter-clockwise quarter turns.  The value may be negative.
\fBepstoimg\fR searches the input file for the first valid
\fB%%Orientation:\fR or \fB%%PageOrientation:\fR comment to set the
orientation of the page; the value of this option is applied relative
to that default.

The width and height of the output, if
specified, are applied after rotation (so \fB\-w 600\fR produces a
600-pixel wide image regardless of the page's orientation).
.TP
\fB\-s\fR, \fB\-\-resolution=\fR<res>
This option controls the resolution (in DPI) of the Ghostscript
interpreter.  Normally \fBepstoimg\fR sets this automatically based on
the desired output size and input bounding box, if any, but this
option can be used to override the default value.  The resolution may
be a single number, or two numbers joined by a letter 'x' to specify
different resolutions in X and Y.  Low values will produce
fast, lower-quality output; higher values may result in nicer output
at the expense of processing time.
.TP
\fB\-g\fR, \fB\-\-grayscale\fR
Converts the image to grayscale before saving.
.TP
\fB\-o\fR, \fB\-\-output=\fR<filename>
Sets the output filename.  The default is to write to stdout.  At
least one of \fB\-o\fR or \fB\-f\fR must be given.  If only \fB\-o\fR
is given, the output format is determined from the file extension.

If the input contains multiple pages, then a four-digit page number is
inserted before the extension.  Thus, "\-o foo.png" will actually
produce output files "foo.0001.png", "foo.0002.png", etc.  If the
input is just a single page, the output filename is used unchanged.
.TP
\fB\-f\fR, \fB\-\-format\fR=\fIFMT\fR
Sets the output image format.  Valid values are: "BMP", "GIF", "JPEG", "PCX",
"PNG", "PPM", "TIFF".  This overrides the format inferred from the
filename extension given with \fB\-o\fR, if any.
.TP
\fB\-i\fR, \fB\-\-interpreter=\fR<gs>
This specifies the location of the Ghostscript interpreter.  If the
first character is a '+', then the directories in the PATH environment variable are
searched for the remainder of the string and the first match is used.
The default value is set at install time.
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Normally \fBepstoimg\fR prints a message to stderr as each page of the
input file is processed.  \fB\-q\fR suppresses all output to stderr,
except for actual error messages.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
The opposite of \fB\-q\fR, this causes additional information to be
printed to stderr as the program runs.
.SH COPYRIGHT
Copyright \(co 2003 Doug Zongker
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.
.PP
PostScript is a registered trademark of Adobe Systems Incorporated.



CHANGES
version 1.0    :  4 Sep 2003

   PIL's handling of paletted images changed in version 1.1.4; imgtops
   now handles both the new and older versions.  Haven't run across
   any major bugs since March, so we'll call it 1.0!
	
version 0.9.2  :  24 Mar 2003

   imgtops handles most PIL image modes via conversion.  (Notably,
   RGBA images are handled by dropping the alpha channel.)

   ASCII data line lengths are limited to 75 characters by default.
   Added -n option to change this.

   Level 3 output files will now print on level 2 devices, but they
   will show a grey box and a numeral '3' in place of the image.

version 0.9.1  :   2 Mar 2003

   Added rotation option to epstoimg.  Changed short form of
   "resolution" option to "-s".

version 0.9    :  22 Feb 2003

   First released version.


imgtopslib/epstoimg.py
import os, tempfile, sys
import getopt
import Image
from dimensions import interpret_dimension
import re
import cStringIO

version = 'epstoimg 1.0'
copyright = '''Copyright (C) 2003 Doug Zongker
This is free software; see the accompanying license for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.'''

err = sys.stderr

output_formats = ('BMP', 'GIF', 'JPEG', 'PCX', 'PNG', 'PPM', 'TIFF')
extension_map = { '.bmp' : 'BMP', '.dib' : 'BMP',
                  '.gif' : 'GIF',
                  '.jpeg' : 'JPEG', '.jpe' : 'JPEG', '.jpg' : 'JPEG',
                  '.pcx' : 'PCX',
                  '.png' : 'PNG',
                  '.ppm' : 'PPM', '.pgm' : 'PPM', '.pbm' : 'PPM',
                  '.tiff' : 'TIFF', '.tif' : 'TIFF',
                  }
                  

bbox_re = re.compile( r'^%%BoundingBox:\s*(-?\d+)\s+(-?\d+)\s+(-?\d+)\s+(-?\d+)' )
orient_re = re.compile( r'^%%(?:Page)?Orientation:\s*([a-zA-Z]+)' )
pagesize_re = re.compile( r'PageSize \[([\d\.]+) ([\d\.]+)\]' )

class Parameters:
    def __init__( self ):
        self.verbose = 10
        self.grayscale = 0
        
        self.margin = 0
        self.padding = 0
        self.paddingcolor = (255, 255, 255)
        self.res = None

        self.width = None
        self.height = None

        self.output = None
        self.format = None

        self.turns = 0

        self.gs = '+gs'

    def finish( self ):
        if self.gs and self.gs[0] == '+':
            self.found_gs = self.find_program( self.gs[1:] )
        else:
            self.found_gs = os.path.abspath( self.gs )
        if not self.found_gs or not os.path.isfile(self.found_gs) or \
               not os.access(self.found_gs,os.X_OK):
            print >> err, 'Ghostscript interpreter "%s" not found' % (self.gs,)
            sys.exit(1)
        if params.verbose >= 20:
            print >> err, 'Ghostscript is [%s]' % (self.found_gs,)

    def find_program( self, progname ):
        try:
            paths = os.environ['PATH'].split( os.pathsep )
        except KeyError:
            paths = ['']
        for path in paths:
            p = os.path.join( path, progname )
            if os.path.exists( p ):
                return os.path.abspath( p )

        print >> err, '"%s" not found in PATH' % (progname,)
        return None

params = Parameters()


def read_headers( fn ):
    bfound = 0
    ofound = 0
    bbox = None
    orient = 0
    
    f = open( fn )
    for i in f:
        if not bfound:
            m = bbox_re.match( i )
            if m:
                bbox = tuple([int(i) for i in m.groups()])
                bfound = 1
        if not ofound:
            m = orient_re.match( i )
            if m:
                o = m.groups(1)[0].lower()
                if o == 'portrait':
                    orient = 0
                    ofound = 1
                elif o == 'landscape':
                    orient = 3
                    ofound = 1
                else:
                    print >> err, 'skipping bad DSC comment: %r' % (i.rstrip(),)
        if bfound and ofound:
            break
                
    if params.verbose >= 20:
        print >> err, 'bbox is %r\norient is %d' % (bbox,orient)
    return bbox, orient

def apply_margin( bbox ):
    m = params.margin
    if m < 0 and (bbox[2]-bbox[0] <= 2*m or bbox[3]-bbox[1] <= 2*m):
        print >> err, "negative margin leaves bbox empty!"
        sys.exit(1)
    return bbox[0]-m, bbox[1]-m, bbox[2]+m, bbox[3]+m

def compute_resolution( bbox ):
    if params.res:
        return params.res

    if bbox:
        bw, bh = bbox[2]-bbox[0], bbox[3]-bbox[1]
        w, h = get_desired_size( (bw, bh) )
        rx = 288.0 * w / bw
        ry = 288.0 * h / bh
        return '%sx%s' % (rx, ry)
    else:
        return '300'
    

def process_ps_file( fn, bbox ):
    res = compute_resolution( bbox )
    if params.verbose >= 20:
        print 'Ghostscript resolution is [%s]' % (res,)
    
    tmpdir = tempfile.mktemp()
    os.mkdir( tmpdir )
    page1 = None
    try:
        cmd = params.found_gs + ' -sDEVICE=ppmraw -sOutputFile=%s/page%%08d.ppm -r%s' % (tmpdir, res)

        to_gs, from_gs = os.popen2( cmd )

        to_gs.write( '<< ' )
        if bbox:
            to_gs.write( '/PageSize [ %f %f ] /BeginPage { pop %f %f translate } ' %
                         (bbox[2]-bbox[0], bbox[3]-bbox[1], -bbox[0], -bbox[1]) )
        to_gs.write( '/EndPage { exch pop 2 ne (%stdout) (w) file (\nPageSize ) writestring currentpagedevice /PageSize get == }' )
        to_gs.write( '>> setpagedevice <%s> run quit\n' % (fn.encode('hex'),) )
        to_gs.flush()

        page = 1
        while 1:
            line = from_gs.readline()
            if line == '':
                break
            m = pagesize_re.match( line )
            if m:
                pagesize = tuple([float(i) for i in m.groups()])
                
            elif line.find( 'press <return> to continue' ) != -1:
                imagefile = '%s/page%08d.ppm' % (tmpdir, page)

                if params.output:
                    if page == 1:
                        page1 = cStringIO.StringIO()
                        f = page1
                    else:
                        root, ext = os.path.splitext( params.output )
                        outfn = root + '.%04d' % (page,) + ext
                        f = open( outfn, 'wb' )
                else:
                    buffer = cStringIO.StringIO()
                    f = buffer
                        
                process_image( imagefile, bbox, pagesize, page, f )
                os.remove( imagefile )

                if not params.output:
                    sys.stdout.write( buffer.getvalue() )
                    buffer.close()
                    if params.verbose >= 20:
                        print >> err, '  wrote to <stdout>'
                elif page > 1:
                    f.close()
                    if params.verbose >= 10:
                        print >> err, 'wrote page %d to [%s]' % (page, outfn)
                
                page += 1
                try:
                    to_gs.write( '\n' )
                    to_gs.flush()
                except IOError, errno:
                    if sys.platform == 'win32' and errno.errno == 22:
                        pass
                    else:
                        raise
    finally:
        for i in os.listdir( tmpdir ):
            os.remove( os.path.join(tmpdir, i) )
        os.rmdir( tmpdir )

    if page1:
        if page > 2:
            root, ext = os.path.splitext( params.output )
            outfn = root + '.0001' + ext
        else:
            outfn = params.output

        f = open( outfn, 'wb' )
        f.write( page1.getvalue() )
        f.close()
        page1.close()
    
        if params.verbose >= 10:
            print >> err, 'wrote page 1 to [%s]' % (outfn,)

def process_image( fn, bbox, pagesize, pagenum, f ):
    if params.verbose >= 10:
        print >> err, 'processing page %d...' % (pagenum,)
        
    if params.verbose >= 20:
        print >> err, '  temp image is [%s]' % (fn,)
        print >> err, '  page size: %.3f %.3f' % pagesize
        
    newsize = get_desired_size( pagesize )

    im = Image.open( fn )
    if im.size != newsize:
        if params.verbose >= 20:
            print >> err, '  reducing size from %r to %r' % (im.size, newsize)
        im = im.resize( newsize, Image.ANTIALIAS )
    if params.grayscale and im.mode != 'L':
        if params.verbose >= 20:
            print >> err, '  converting to grayscale'
        im = im.convert( 'L' )
    if params.turns:
        if params.verbose >= 20:
            print >> err, '  rotating %d quarter-turns' % (params.turns,)
        im = im.transpose( (Image.ROTATE_90,Image.ROTATE_180,Image.ROTATE_270)[params.turns-1] )
    if params.padding > 0:
        if params.verbose >= 20:
            print >> err, '  padding image'
        p = params.padding
        im2 = Image.new( 'RGB', (newsize[0]+2*p, newsize[1]+2*p), params.paddingcolor )
        im2.paste( im, (p,p) )
        im = im2

    if params.verbose >= 20:
        print >> err, '  writing in %s format' % (params.format,)
    im.save( f, params.format )


def get_desired_size( pagesize ):
    w = params.width
    h = params.height
    p = 2 * params.padding

    if params.turns % 2:
        w, h = h, w
        
    if w and h:
        w -= p
        h -= p
    elif w:
        w -= p
        h = int( float(w) / pagesize[0] * pagesize[1] ) 
    elif h:
        h -= p
        w = int( float(h) / pagesize[1] * pagesize[0] )
    else:
        w = int( pagesize[0] )
        h = int( pagesize[1] )

    return w, h
        
        


def usage( gs ):
    print 'usage: %s [options] <filename>' % (sys.argv[0],)
    print 
    print '  -?, --help                   show this message'
    print '  --version                    print version info and exit'
    print 
    print '  -w, --width=#                set output image width (in pixels)'
    print '  -h, --height=#               set output image height (in pixels)'
    print '  -m, --margin=<dim>           set bbox margin (default none)'
    print '        example dimensions:      4in, 35mm, 30.1cm, 144pt'
    print '  -p, --padding=#              set image padding (in pixels; default=0)'
    print '  -b, --black                  pad with black instead of white'
    print '  -r, --rotation=#             integer # of quarter-turns to rotate (ccw)'
    print '  -s, --resolution=<res>       set GS resolution manually'
    print '  -g, --grayscale              output grayscale image'
    print
    print '  -o, --output=<filename>      output file name'
    print '  -f, --format=FMT             output image format (overrides extension)'
    print '          formats: ', ', '.join(output_formats)
    print
    print '  -i, --interpreter=<gs>       location of Ghostscript'
    print '          prefix with "+" to search PATH'
    print '          default: "%s"' % (gs,)
    print '  -q, --quiet                  run silently, except for errors'
    print '  -v, --verbose                be verbose'

def showversion():
    print version
    print
    print copyright
    
        
def main( **kw ):
    for k, v in kw.iteritems():
        setattr( params, k, v )
    
    if len(sys.argv) < 2:
        usage( params.gs )
        return 2

    try:
        opts, args = getopt.getopt(sys.argv[1:], '?w:h:m:p:bo:qvgf:r:s:i:',
                                   ['width=',
                                    'height=',
                                    'margin=',
                                    'padding=',
                                    'black',
                                    'output=',
                                    'quiet',
                                    'verbose',
                                    'grayscale',
                                    'format=',
                                    'rotation=',
                                    'resolution=',
                                    'interpreter=',
                                    'version',
                                    ])
    except getopt.GetoptError:
        usage( params.gs )
        return 2
    output = None
    for o, a in opts:
        if o in ('-?', '--help'):
            usage( params.gs )
            return 0
        elif o == '--version':
            showversion()
            return 0
        elif o in ('-v', '--verbose'):
            params.verbose = 20
        elif o in ('-q', '--quiet'):
            params.verbose = 0
        elif o in ('-g', '--grayscale'):
            params.grayscale = 1
        elif o in ('-b', '--black'):
            params.paddingcolor = (0,0,0)
        elif o in ('-w', '--width'):
            try:
                params.width = int(a)
            except ValueError:
                print >> err, "didn't understand width value %r" % (a,)
                return 1
        elif o in ('-h', '--height'):
            try:
                params.height = int(a)
            except ValueError:
                print >> err, "didn't understand height value %r" % (a,)
                return 1
        elif o in ('-p', '--padding'):
            try:
                params.padding = int(a)
            except ValueError:
                print >> err, "didn't understand padding value %r" % (a,)
                return 1
        elif o in ('-m', '--margin'):
            params.margin = interpret_dimension(a)
            if params.margin is None:
                print >> err, "didn't understand margin dimension %r" % (a,)
                return 1
        elif o in ('-o', '--output'):
            params.output = a
        elif o in ('-f', '--format'):
            if a.upper() not in output_formats:
                print >> err, "output format %r is unknown" % (a,)
                return 1
            params.format = a.upper()
        elif o in ('-r', '--rotation'):
            try:
                params.turns = int(a)
            except ValueError:
                print >> err, "bad value %r for rotation, must be integer" % (a,)
                return 1
        elif o in ('-s', '--resolution'):
            nums = a.lower().split('x')
            if len(nums) != 1 and len(nums) != 2:
                print >> err, "didn't understand resolution %r" % (a,)
                return 1
            try:
                nums = [str(float(i)) for i in nums]
            except ValueError:
                print >> err, "didn't understand resolution %r" % (a,)
                return 1
            params.res = 'x'.join(nums)
        elif o in ('-i', '--interpreter'):
            params.gs = a

    params.finish()
            

    if params.width is not None:
        if params.width <= 0:
            print >> err, "width must be positive"
            return 1
        if params.width - 2 * params.padding <= 0:
            print >> err, "padding too big for width"

    if params.height is not None:
        if params.height <= 0:
            print >> err, "height must be positive"
            return 1
        if params.height - 2 * params.padding <= 0:
            print >> err, "padding too big for height"

    if len(args) == 0:
        print >> err, "no input file specified!"
        return 1
    
    if len(args) > 1:
        print >> err, "too many input files specified!"
        return 1

    if params.output is None:
        if params.format is None:
            print >> err, "must specify output filename and/or format!"
            return 1
        
        if sys.platform == "win32":
            import msvcrt
            msvcrt.setmode(sys.stdout.fileno(), os.O_BINARY)
    else:
        if params.format is None:
            root, ext = os.path.splitext( params.output )
            try:
                params.format = extension_map[ext.lower()]
            except KeyError:
                print >> err, "can't determine output format from file extension"
                return 1
        

    user_ps = args[0]
    user_path = os.path.abspath( user_ps )
    bbox, orient = read_headers( user_path )
    params.turns = (params.turns + orient) % 4
    if bbox:
        bbox = apply_margin( bbox )
    
    process_ps_file( user_path, bbox )

if __name__ == "__main__":
    r = main()
    sys.exit( r )
    
